h/exceptions.py

```diff
@@ -1,52 +1,46 @@
 # -*- coding: utf-8 -*-
-
 """Exceptions raised by the h application."""
 
 from __future__ import unicode_literals
 
-from h.i18n import TranslationString as _  # noqa: N813
+from pyramid import httpexceptions
 
+from h.i18n import TranslationString as _  # noqa: N813
 
-class APIError(Exception):
 
-    """Base exception for problems handling API requests."""
+class APIError(httpexceptions.HTTPError):
+    """A problem handling an API request."""
 
-    def __init__(self, message, status_code=500):
-        self.status_code = status_code
-        super(APIError, self).__init__(message)
+    def __init__(self, detail, status_code=500):
+        super(APIError, self).__init__(detail, status_code=status_code)
 
 
-class ConflictError(APIError):
-    """
-    Exception raised if client request represents a duplicate of an
-    existing resource.
-    """
+class ConflictError(httpexceptions.HTTPConflict):
+    """An API request represents a duplicate of an existing resource."""
 
-    def __init__(self, message=_("Conflict")):
-        super(ConflictError, self).__init__(message, status_code=409)
+    def __init__(self, detail=None):
+        if detail is None:
+            detail = _("Conflict")
 
+        super(ConflictError, self).__init__(detail)
 
-class OAuthTokenError(APIError):
 
+class OAuthTokenError(httpexceptions.HTTPUnauthorized):
     """
-    Exception raised when an OAuth token request failed.
+    An OAuth token request failure.
 
-    This specifically handles OAuth errors which have a type (``message``) and
-    a description (``description``).
+    Raising an exception of this class causes an error response that has
+    a type (``message``) and a description (``description``).
     """
 
-    def __init__(self, message, type_, status_code=400):
+    def __init__(self, detail, type_):
         self.type = type_
-        super(OAuthTokenError, self).__init__(message, status_code=status_code)
-
+        super(OAuthTokenError, self).__init__(detail)
 
-class PayloadError(APIError):
 
-    """
-    Exception raised for API requests made with missing/invalid
-    payloads.
-    """
+class PayloadError(httpexceptions.HTTPBadRequest):
+    """An API request has a missing or invalid payload."""
 
     def __init__(self):
-        message = _("Expected a valid JSON payload, but none was found!")
-        super(PayloadError, self).__init__(message, status_code=400)
+        detail = _("Expected a valid JSON payload, but none was found!")
+        super(PayloadError, self).__init__(detail)
```

Comment:
This is an example of a class that actually adds something to the `HTTPBadRequest` class that it subclasses -- it adds a useful custom `detail` string.